chainer/functions/normalization/batch_normalization.py

```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
+            libcudnn.deriveBNTensorDescriptor(derivedBnDesc.value,
+                                              x_desc.value, self.mode)
+            one = numpy.array(1, dtype=dtype).ctypes
+            zero = numpy.array(0, dtype=dtype).ctypes
+            gx = cuda.cupy.empty_like(x)
+            ggamma = cuda.cupy.empty_like(gamma)
+            gbeta = cuda.cupy.empty_like(gamma)
+            libcudnn.batchNormalizationBackward(
+                handle, self.mode, one.data, zero.data,
+                one.data, zero.data, x_desc.value, x.data.ptr,
+                x_desc.value, gy.data.ptr, x_desc.value, gx.data.ptr,
+                derivedBnDesc.value, gamma.data.ptr,
+                ggamma.data.ptr, gbeta.data.ptr,
+                self.eps, self.mean_cache.data.ptr, self.var_cache.data.ptr)
         else:
-            inv_m = numpy.float32(1) / m
-            gx = cuda.elementwise(
-                'T gy, T x_hat, T gamma, T std, T ggamma, T gbeta, T inv_m',
-                'T gx',
-                'gx = (gamma / std) * (gy - (x_hat * ggamma + gbeta) * inv_m)',
-                'bn_bwd')(gy, self.x_hat, gamma[expander], self.std[expander],
-                          ggamma[expander], gbeta[expander], inv_m)
+            gbeta = gy.sum(axis=axis)
+            if self.train:
+                ggamma = (gy * self.x_hat).sum(axis=axis)
+                inv_m = numpy.float32(1) / m
+                gx = cuda.elementwise(
+                    'T gy, T x_hat, T gamma, T std, T ggamma, T gbeta, \
+                    T inv_m',
+                    'T gx',
+                    'gx = (gamma / std) * (gy - (x_hat * ggamma + gbeta) * \
+                    inv_m)',
+                    'bn_bwd')(gy, self.x_hat, gamma[expander],
+                              self.std[expander],
+                              ggamma[expander], gbeta[expander], inv_m)
+            else:
+                # Note: Under normal conditions, this will never be executed
+                # because fixed-mean-variance mode is normally
+                # only used at test/evaluation time.
+                # If cuDNN was called on the forward call, self.x_hat will
+                # be None.
+                if self.x_hat is None:
+                    self.std = xp.sqrt(self.fixed_var,
+                                       dtype=self.fixed_var.dtype)
+                    x_mu = x - self.fixed_mean[expander]
+                    self.x_hat = x_mu / self.std[expander]
+                ggamma = (gy * self.x_hat).sum(axis=axis)
+                gx = cuda.elementwise(
+                    'T gy, T x_hat, T gamma, T std',
+                    'T gx',
+                    'gx = (gamma / std) * gy',
+                    'bn_bwd')(gy, self.x_hat, gamma[expander],
+                              self.std[expander])
         return gx, ggamma, gbeta
 
 
-def batch_normalization(x, gamma, beta, eps=1e-5):
+def batch_normalization(x, gamma, beta, running_mean=None,
```

Comment:
Thank you for the comment. I was just made aware that we should try to maintain backward compatibility on any documented functions, even if they will probably only ever be used by the corresponding link. I will try to rewrite to maintain backward compatibility.